src/snowflake/connector/cache.py

```diff
@@ -0,0 +1,264 @@
+#
+# Copyright (c) 2012-2021 Snowflake Computing Inc. All rights reserved.
+#
+
+from __future__ import annotations
+
+import datetime
+from collections.abc import Iterator
+from threading import Lock
+from typing import Generic, TypeVar
+
+from typing_extensions import NamedTuple, Self
+
+now = datetime.datetime.now
+
+T = TypeVar("T")
+
+
+class CacheEntry(NamedTuple, Generic[T]):
+    expiry: datetime.datetime
+    entry: T
+
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+def is_expired(d: datetime.datetime) -> bool:
+    return now() >= d
+
+
+class SFDictCache(Generic[K, V]):
+    """A generic in-memory cache that acts somewhat like a dictionary.
+
+    Notes:
+    - Unlike normal dictionaries keys, values and items return list
+      materialized at call time, instead of returning a view object.
+    """
+
+    def __init__(
+        self,
+        entry_lifetime: int = 60 * 60 * 24,  # 1 day in seconds
+    ) -> None:
+        self._entry_lifetime = datetime.timedelta(seconds=entry_lifetime)
+        self._cache: dict[K, CacheEntry[V]] = {}
+        self._lock = Lock()
+        self._reset_telemetry()
+
+    @classmethod
+    def from_dict(
+        cls,
+        _dict: dict[K, V],
+        **kw,
+    ) -> Self:
+        cache = cls(**kw)
+        for k, v in _dict.items():
+            cache[k] = v
```

Comment:
I feel we may want to `deepcopy` k and v here to keep an internal copy for the SFDictCache.

otherwise we have to be aware that the SFDictCache is keeping the same reference as the dict derived from

Reply:
sounds reasonable to me, I don't have strong opinions on deepcopy the objects. As long as we're aware of the "same references" here and know how to deal with it then we are good.